brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java

```diff
@@ -0,0 +1,127 @@
+
+    this.validationPeriod = validationPeriod;
+  }
+
+  /**
+   * Updates BRVS whitelist according to *client whitelist* - if client added an address -> add the
+   * address to BRVS whitelist with new validation time - if client removed an address -> remove it
+   * from BRVS whitelist
+   *
+   * @param transaction Iroha proto transaction
+   * @return true
+   */
+  @Override
+  public boolean isSatisfiedBy(TransactionOuterClass.Transaction transaction) {
+    String clientId = transaction.getPayload().getReducedPayload().getCreatorAccountId();
+
+    return transaction
+        .getPayload()
+        .getReducedPayload()
+        .getCommandsList()
+        .stream()
+        .filter(Command::hasSetAccountDetail)
+        .map(Command::getSetAccountDetail)
+        .filter(cmd -> cmd.getAccountId().equals(clientId)
+            && (cmd.getKey().equals(WhitelistUtils.ETH_WHITELIST_KEY)
+            || cmd.getKey().equals(WhitelistUtils.BTC_WHITELIST_KEY)))
+        .allMatch(tx -> {
+          try {
+            String whitelistKey = tx.getKey();
+            List<String> clientWhitelist = WhitelistUtils.deserializeClientWhitelist(tx.getValue());
+            logger.info("Client " + clientId + " changed whitelist " + whitelistKey + " to "
+                + clientWhitelist);
+
+            // get old whitelist that was set by BRVS as Pairs(address -> validation_time)
+            Map<String, Long> oldWhitelistValidated = WhitelistUtils.getBRVSWhitelist(
+                brvsAccountId,
+                brvsAccountKeyPair,
+                irohaAPI,
+                clientId,
+                whitelistKey
+            );
+
+            // When whitelist is validated
+            long validationTime = System.currentTimeMillis() / 1000 + validationPeriod;
+            logger.info("ValidationTime: " + validationTime);
+
+            // Prepare new whitelist
+            // remove old addresses that user has removed
+            Map<String, Long> newWhitelistValidated = oldWhitelistValidated.entrySet().stream()
+                .filter(address -> clientWhitelist.contains(address.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            // add new from user whitelist with new validationTime
+            newWhitelistValidated.putAll(
+                clientWhitelist.stream()
+                    .filter(address -> !oldWhitelistValidated.containsKey(address))
+                    .collect(Collectors.toMap(a -> a, t -> validationTime))
+            );
+
+            if (newWhitelistValidated.equals(oldWhitelistValidated)) {
+              logger.info("No changes in whitelist, nothing to update");
+            } else {
+              String jsonFin = WhitelistUtils.serializeBRVSWhitelist(newWhitelistValidated);
+              logger.info("Send whitelist to Iroha: " + jsonFin);
+
+              irohaAPI.transactionSync(
+                  Transaction.builder(brvsAccountId)
```

Comment:
Please provide `createdTime` equal to the beginning of the day in order to support many BRVS instances in the domain so brvs account could be MST

Reply:
It is a bad approach to set it equals to the beginning of the day. What if BRVS discovers tx at midnight? Part of multisig transactions will have one date and part will have another. It is better to take time from external source, for example incoming transaction.

Reply:
Sure you can do it in a such way. My point was just about time sync, not exact approach 👍

Reply:
done